internal/service/datazone/project.go

```diff
@@ -0,0 +1,452 @@
+		DomainIdentifier: aws.String(plan.DomainId.ValueString()),
+	}
+	if !plan.Description.IsNull() {
+		in.Description = aws.String(plan.Description.ValueString())
+	}
+	if !plan.GlossaryTerms.IsNull() {
+		in.GlossaryTerms = aws.ToStringSlice(flex.ExpandFrameworkStringList(ctx, plan.GlossaryTerms))
+	}
+
+	out, err := conn.CreateProject(ctx, in)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			create.ProblemStandardMessage(names.DataZone, create.ErrActionCreating, ResNameProject, plan.Name.String(), err),
+			err.Error(),
+		)
+		return
+	}
+	if out == nil {
+		resp.Diagnostics.AddError(
+			create.ProblemStandardMessage(names.DataZone, create.ErrActionCreating, ResNameProject, plan.Name.String(), nil),
+			errors.New("failure when creating").Error(),
+		)
+		return
+	}
+	if !(out.FailureReasons == nil) && len(out.FailureReasons) > 0 {
+		for _, x := range out.FailureReasons {
+			resp.Diagnostics.AddError(
+				create.ProblemStandardMessage(names.DataZone, create.ErrActionCreating, ResNameProject, plan.Name.String(), nil),
+				errors.New("error message: "+*x.Message+" error code: "+*x.Code).Error(),
+			)
+		}
+		return
+	}
+
+	resp.Diagnostics.Append(flex.Flatten(ctx, out, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	createTimeout := r.CreateTimeout(ctx, plan.Timeouts)
+	_, err = waitProjectCreated(ctx, conn, plan.DomainId.ValueString(), plan.ID.ValueString(), createTimeout)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			create.ProblemStandardMessage(names.DataZone, create.ErrActionWaitingForCreation, ResNameProject, plan.Name.String(), err),
+			err.Error(),
+		)
+		return
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
+}
+
+func (r *resourceProject) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	conn := r.Meta().DataZoneClient(ctx)
+	var state resourceProjectData
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	//out, err := conn.GetProject(ctx, in)
+	out, err := findProjectByID(ctx, conn, state.DomainId.ValueString(), state.ID.ValueString())
+	if tfresource.NotFound(err) {
+		resp.State.RemoveResource(ctx)
+		return
+	}
+
+	if err != nil {
+		resp.Diagnostics.AddError(
+			create.ProblemStandardMessage(names.DataZone, create.ErrActionSetting, ResNameProject, state.ID.String(), err),
+			err.Error(),
+		)
+		return
+	}
+	resp.Diagnostics.Append(flex.Flatten(ctx, out, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+}
+
+func (r *resourceProject) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	conn := r.Meta().DataZoneClient(ctx)
+
+	var plan, state resourceProjectData
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	if plan.DomainId != state.DomainId {
```

Comment:
the Plugin Framework provides builtin functions for checking equality: `plan.DomainId.Equals(state.DomainId)` will do equality checks for both `Null` and `Unknown` values.

We should also try to use `flex.Expand` here as well. `internal/services/opensearchserverless/access_policy.go` has a pretty good pattern for doing these equality checks, along with autoFlex.